ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/ConflictCheckingLedgerBridge.scala

```diff
@@ -0,0 +1,623 @@
+          case (KeyActive(expected), actual) if !actual.contains(expected) =>
+            Left(InconsistentContractKey(Some(expected), actual)(transaction))
+          case _ => Right(())
+        }
+      }
+
+  override def close(): Unit = {
+    logger.info("Shutting down ConflictCheckingLedgerBridge.")
+    queue.complete()
+    conflictCheckingQueue.complete()
+  }
+}
+
+object ConflictCheckingLedgerBridge {
+  private implicit val errorLoggingContext: ContextualizedErrorLogger = NoLogging
+
+  trait Submission extends Serializable with Product
+  object Submission {
+    case class Transaction(
+        submitterInfo: SubmitterInfo,
+        transactionMeta: TransactionMeta,
+        transaction: SubmittedTransaction,
+        estimatedInterpretationCost: Long,
+        blindingInfo: BlindingInfo,
+    ) extends Submission
+    case class Config(
+        maxRecordTime: Time.Timestamp,
+        submissionId: Ref.SubmissionId,
+        config: Configuration,
+    ) extends Submission
+    case class AllocateParty(
+        hint: Option[Ref.Party],
+        displayName: Option[String],
+        submissionId: Ref.SubmissionId,
+    ) extends Submission
+
+    case class UploadPackages(
+        submissionId: Ref.SubmissionId,
+        archives: List[Archive],
+        sourceDescription: Option[String],
+    ) extends Submission
+
+    case class Rejection(rejection: SoxRejection) extends Submission
+  }
+
+  private[this] val logger = ContextualizedLogger.get(getClass)
+
+  def successMapper(submission: Submission, index: Long, participantId: Ref.ParticipantId): Update =
+    submission match {
+      case s: Submission.AllocateParty =>
+        val party = s.hint.getOrElse(UUID.randomUUID().toString)
+        Update.PartyAddedToParticipant(
+          party = Ref.Party.assertFromString(party),
+          displayName = s.displayName.getOrElse(party),
+          participantId = participantId,
+          recordTime = Time.Timestamp.now(),
+          submissionId = Some(s.submissionId),
+        )
+
+      case s: Submission.Config =>
+        Update.ConfigurationChanged(
+          recordTime = Time.Timestamp.now(),
+          submissionId = s.submissionId,
+          participantId = participantId,
+          newConfiguration = s.config,
+        )
+
+      case s: Submission.UploadPackages =>
+        Update.PublicPackageUpload(
+          archives = s.archives,
+          sourceDescription = s.sourceDescription,
+          recordTime = Time.Timestamp.now(),
+          submissionId = Some(s.submissionId),
+        )
+
+      case s: Submission.Transaction =>
+        Update.TransactionAccepted(
+          optCompletionInfo = Some(s.submitterInfo.toCompletionInfo),
+          transactionMeta = s.transactionMeta,
+          transaction = s.transaction.asInstanceOf[CommittedTransaction],
+          transactionId = Ref.TransactionId.assertFromString(index.toString),
+          recordTime = Time.Timestamp.now(),
```

Comment:
for correctness: this recordTime needs to be checked against the time-model, not the now() as the submission enters the parallel conflict checking stage
(giving the devil it's due: if this is really low latency and high TP then probability is fairly low to cause this an issue)